raiden/messages.py

```diff
@@ -595,12 +642,24 @@ def from_event(cls, event):
             signature=EMPTY_SIGNATURE,
         )
 
+    def _data_to_sign(self) -> bytes:
+        return pack_data(
+            # TODO: should we add cmdid?
```

Comment:
Hummm, so this is a trick one. Nope, we cannot add the `cmdid` here.

This function returns the data that will be signed, this message has only one signature that is used for a few things:

1. The most important is to have a signature that can be used by the smart contract, so it must follow the exact same format.
2. This is also used to authenticate the data for the off-chain recipient, which proves to it that this is a valid request.

So, the smart contract currently uses [this format](https://github.com/raiden-network/raiden-contracts/blob/1a543f5fce48603f25de831f9abd47f66848fb5f/raiden_contracts/data/source/raiden/TokenNetwork.sol#L1558-L1568), note unlike the [balance proof](https://github.com/raiden-network/raiden-contracts/blob/1a543f5fce48603f25de831f9abd47f66848fb5f/raiden_contracts/data/source/raiden/TokenNetwork.sol#L1476) this does not have an `additional_hash` field, this means the signature has to have exactly these fields.

Even if we had an `additional_hash` field, it would not work for the `WithdrawConfirmation` message, because the messages are not identical and the smart contract current expects both signatures [1](https://github.com/raiden-network/raiden-contracts/blob/1a543f5fce48603f25de831f9abd47f66848fb5f/raiden_contracts/data/source/raiden/TokenNetwork.sol#L400-L406) [2](https://github.com/raiden-network/raiden-contracts/blob/1a543f5fce48603f25de831f9abd47f66848fb5f/raiden_contracts/data/source/raiden/TokenNetwork.sol#L407-L413) to use the exact same data.

To support this we have to change the smart contract code, I can think of two options:

1. Have two different `MessageTypeID`s, one for the request and another for the confirmation, and then make the cmdid equal to the messageid.
2. Add an additional hash to the `Withdraw`, just like the balance proof. This would then be signed by the partner, and the sender would sign the whole thing. The data would then look something like `tn_address || chain_id || Withdraw || channel_id || participant || total_withdraw|| additional_hash || partner_signature || participant_signature`.

I talked to @pirapira about this.

Reply:
See https://github.com/raiden-network/raiden-contracts/issues/1110.